src/python/tensorflow_cloud/core/validate.py

```diff
@@ -73,83 +64,113 @@ def validate(
+):
     """Validates cluster config params."""
     if not isinstance(chief_config, machine_config.MachineConfig):
-        raise ValueError("Invalid `chief_config` input. "
-                         'Expected "auto" or `MachineConfig` instance. '
-                         "Received {}.".format(chief_config))
+        raise ValueError(
+            "Invalid `chief_config` input. "
+            'Expected "auto" or `MachineConfig` instance. '
+            "Received {}.".format(chief_config)
+        )
 
     if worker_count < 0:
-        raise ValueError("Invalid `worker_count` input. "
-                         "Expected a postive integer value. "
-                         "Received {}.".format(worker_count))
+        raise ValueError(
+            "Invalid `worker_count` input. "
+            "Expected a postive integer value. "
+            "Received {}.".format(worker_count)
+        )
 
-    if worker_count > 0 and not isinstance(worker_config,
-                                           machine_config.MachineConfig):
-        raise ValueError("Invalid `worker_config` input. "
-                         'Expected "auto" or `MachineConfig` instance. '
-                         "Received {}.".format(worker_config))
+    if worker_count > 0 and not isinstance(worker_config, machine_config.MachineConfig):
+        raise ValueError(
+            "Invalid `worker_config` input. "
+            'Expected "auto" or `MachineConfig` instance. '
+            "Received {}.".format(worker_config)
+        )
 
     if machine_config.is_tpu_config(chief_config):
-        raise ValueError("Invalid `chief_config` input. "
-                         "`chief_config` cannot be a TPU config. "
-                         "Received {}.".format(chief_config))
+        raise ValueError(
+            "Invalid `chief_config` input. "
+            "`chief_config` cannot be a TPU config. "
+            "Received {}.".format(chief_config)
+        )
 
     if machine_config.is_tpu_config(worker_config):
         if worker_count != 1:
-            raise ValueError("Invalid `worker_count` input. "
-                             "Expected worker_count=1 for TPU `worker_config`. "
-                             "Received {}.".format(worker_count))
-        elif VERSION >= "2.2.0":
-            raise NotImplementedError(
-                "TPUs are only supported for TF version <= 2.1.0")
+            raise ValueError(
+                "Invalid `worker_count` input. "
+                "Expected worker_count=1 for TPU `worker_config`. "
+                "Received {}.".format(worker_count)
+            )
+        elif docker_base_image is None:
+            # If the user has not provided a custom docker image, then verify
+            # that the TF version is compatible with Cloud TPU support.
+            # https://cloud.google.com/ai-platform/training/docs/runtime-version-list#tpu-support
+            version = tf_utils.get_version()
+            if version and version >= "2.2.0":
```

Comment:
Recommend defining this as a constant in tf_utils.py.

This is basically defining the bound of TF version where TPU can be supported.

Maybe defining it as 

TPU_SUPPORTED_TF_VERSION = ["2.1", "2.x", ..."] and explicitly match against it. We expect this list will change over time.

Reply:
(as of today, it is ["1.13", "1.14", "1.15", "2.1"], but missing "2.0", dependency.py defines TF dependency as >=1.15,<3, so ,it should cover ["1.15", "2.1"]

Reply:
Thought about this, since this is not a TF constraint and more of a CAIP constraint, added a helper to the gcp module. At this time i am not sure we should support TF versions < 2.

Reply:
FWIW, TF 1.15 is marked as LTS (Long Term Support) as TensorFlow Enterprise on CAIP. 

Is tfc.core supposed to work with TF 1.15 code at all, given it injects dist-strat code into it? It's fair to completely drop support of it. If so, it should be done at upfront at the beginning of trc.run().

Reply:
(it doesn't have to be in the scope of this PR, so approving)